src/endpoint.rs

```diff
@@ -13,90 +13,174 @@ use message::{Message, Notification, Request};
 use message::Response as MsgPackResponse;
 use codec::Codec;
 
+/// A channel for returning responses from a service.
+///
+/// See [`Service::handle_request`] for more details.
+#[derive(Debug)]
+pub struct ReturnChannel {
+    id: u32,
+    sender: mpsc::UnboundedSender<(u32, Result<Value, Value>)>,
+}
+
+impl ReturnChannel {
+    /// Sends a response back along the return channel.
+    ///
+    /// Each return channel can only be used once, so this consumes the channel.
+    pub fn send(self, result: Result<Value, Value>) {
+        // If there is an error sending, it means that the service has stopped. There's no point in
+        // signalling an error because it's impossible anyway to guarantee that the response got
+        // back to the client (unless you ask the client to send a confirmation).
+        let _ = self.sender.unbounded_send((self.id, result));
+    }
+}
+
 /// The `Service` trait defines how a `MessagePack-RPC` server handles requests and notifications.
 pub trait Service {
-    type Error: Error;
-    type T: Into<Value>;
-    type E: Into<Value>;
+    /// Handle a `MessagePack-RPC` request.
+    ///
+    /// The name of the request is `method`, and the parameters are given in `params`. The response
+    /// to the request should be sent along `return_channel`.
+    ///
+    /// Note that this method is called synchronously within the main event loop, and so it should
+    /// return quickly. If you need to run a longer computation, spawn a new task and give it the
+    /// return channel.
+    ///
+    /// This return channel business might not be the best way of dealing with the issue. A
+    /// previous version had `handle_request` return a Future instead. That was a little awkward
+    /// because it complicated the easy case where the response is immediately available. This
+    /// "return channel" method also has the (potential) advantage of allowing the service to be in
+    /// control of spawning the new task.
+    fn handle_request(&mut self, method: &str, params: &[Value], return_channel: ReturnChannel);
```

Comment:
I'm not sure how I feel about this. On one hand, not returning a future has all the advantages you list, and also has the benefit of simplifying error handling. But on the other hand it's exposing an implementation detail to the users (they have to understand what this `return_channel` does, and use it correctly). But to be honest I lack experience designing rust APIs (I made `rmp-rpc` to learn about Tokio and async programming in the first place), I can't decide which way I prefer. @tylerwhall do you have an opinion about this?

Reply:
So I looked at [tarpc](https://github.com/google/tarpc) yesterday, and they have a nicer approach: the `Service` trait would come with a `type Fut: IntoFuture<Item=Value, Error=Value>` associated type, which is then the return value of all the methods. That way, if you just want to be synchronous then you can set `Fut` to `Result`.

Reply:
That sounds like the best of both world, indeed :)

Now that I take a closer look to this, I realize these changes remove the possiblity to express a failure to handle a request or a notification, but that's fine, because anyway I [currently panic](https://github.com/little-dude/rmp-rpc/blob/0dd5e8e44b2f56ba58725b61f58ecd96cf64fa32/src/endpoint.rs#L70) if `handle_request` or `handle_notification` returns an error, which is way worse.

Reply:
Well, it seemed to me that the only reasonable way to deal with an error from within `handle_request` is either to panic (if the error represents a real fatal bug) or send an error message back to the client (if the error is recoverable). In the panic case, there's no need to bubble the error up before panicking.